completions/bash/brew

```diff
@@ -50,9 +50,13 @@ __brewcomp() {
 # it is too slow and is not worth it just for duplicate elimination.
 __brew_complete_formulae() {
   local cur="${COMP_WORDS[COMP_CWORD]}"
-  local formulas="$(brew search)"
-  local shortnames="$(echo "$formulas" | \grep / | \cut -d / -f 3)"
-  COMPREPLY=($(compgen -W "$formulas $shortnames" -- "$cur"))
+  HOMEBREW_REPOSITORY="${HOMEBREW_REPOSITORY:-$(brew --repo)}"
+
+  local formulae="$(\find $HOMEBREW_REPOSITORY/Library/Taps -type d \( -name Casks -o -name cmd -o -name .github \) -prune -false -o -name '*rb' | \sed 's/\.rb//g' | \sed -E 's/.*\/Taps\/(.*)\/homebrew-/\1\//g' | \sed 's/\/Formula\//\//g')"
```

Comment:
This is great. I think it'd be better handled with a brew command implemented in Bash (e.g. `brew formulae`) that does the same thing (and can be used by e.g. ZSH completions too). How does that sound to you?

Reply:
Sorry for taking so long to reply.  That sounds like a great idea. I'll take a look.

Reply:
Hi, after playing around a bit, it looks like adding it as a proper command (under the cmd directory) also incurs a pretty hefty startup time cost and increases the run time to about ~700ms.

What do you think of doing this?

```diff
diff --git a/Library/Homebrew/brew.sh b/Library/Homebrew/brew.sh
index 814c0ca22..cc615162c 100644
--- a/Library/Homebrew/brew.sh
+++ b/Library/Homebrew/brew.sh
@@ -48,11 +48,15 @@ else
   HOMEBREW_CELLAR="$HOMEBREW_PREFIX/Cellar"
 fi
 
+# shellcheck disable=SC1090
+source "$HOMEBREW_LIBRARY/Homebrew/cmd/formulae.sh"
+
 case "$*" in
-  --prefix)            echo "$HOMEBREW_PREFIX"; exit 0 ;;
-  --cellar)            echo "$HOMEBREW_CELLAR"; exit 0 ;;
-  --repository|--repo) echo "$HOMEBREW_REPOSITORY"; exit 0 ;;
-  --caskroom)          echo "$HOMEBREW_PREFIX/Caskroom"; exit 0 ;;
+  --prefix)               echo "$HOMEBREW_PREFIX"; exit 0 ;;
+  --cellar)               echo "$HOMEBREW_CELLAR"; exit 0 ;;
+  --repository|--repo)    echo "$HOMEBREW_REPOSITORY"; exit 0 ;;
+  --caskroom)             echo "$HOMEBREW_PREFIX/Caskroom"; exit 0 ;;
+  --formulae|--formulas)  homebrew-formulae; exit 0 ;;
 esac
 
 # A depth of 1 means this command was directly invoked by a user.

```

This recovers almost all of the speed.  A --formulae_spec.rb test could even be written that would compare the output to brew search this way.  

I guess it depends on how important that 400ms is to everyone.  This makes the command feel subjectively a lot slower to me, but that's just me.

Reply:
> Hi, after playing around a bit, it looks like adding it as a proper command (under the cmd directory) also incurs a pretty hefty startup time cost and increases the run time to about ~700ms.

Is that with a hot or cold cache? If I repeatedly run `brew shellenv` about 0.5s is removed from the time on the second time it is run.

> What do you think of doing this?

👎🏻 on that, it's a bit of a hack. At most I could see an argument for having a `source`d shared script which is sourced directly in the ZSH completion.

Reply:
@metacolin Actually, you've convinced me 😂. I'm opening a PR to do this for `shellenv` and then you can do the same for `formulae`.

Reply:
https://github.com/Homebrew/brew/pull/8904. Once that's merged you can rebase this PR on there (or merge `master` in) and use the same format I've done there to add a new `formulae` command.

Reply:
@metacollin, for your `sed` commands -- you can use any other separators instead of `/` so that you don't have to escape `/` in patterns. Also, you can combine all three `sed`s into one:

```
\sed -E 's#\.rb##g; s#.*/Taps/(.*)/homebrew-#\1/#g; s#/Formula/#/#g'
```

Then, currently, this returns things like:
<details>

```
homebrew/portable-ruby/Abstract/portable-formula
homebrew/test-bot/spec/spec_helper
homebrew/test-bot/spec/homebrew/step_spec
homebrew/test-bot/spec/homebrew/tests/setup_spec
homebrew/test-bot/spec/stub/formula_installer
homebrew/test-bot/spec/stub/utils/bottles
homebrew/test-bot/spec/stub/tap
homebrew/test-bot/spec/stub/utils
homebrew/test-bot/spec/stub/os
homebrew/test-bot/spec/stub/development_tools
homebrew/test-bot/spec/stub/formula
homebrew/test-bot/spec/global
homebrew/test-bot/lib/test_bot
homebrew/test-bot/lib/tests/cleanup_before
homebrew/test-bot/lib/tests/cleanup_after
homebrew/test-bot/lib/tests/setup
homebrew/test-bot/lib/tests/tap_syntax
homebrew/test-bot/lib/tests/formulae
homebrew/test-bot/lib/step
homebrew/test-bot/lib/test_runner
homebrew/test-bot/lib/test
homebrew/test-bot/lib/test_cleanup
homebrew/test-bot/lib/test_ci_upload
homebrew/bundle/spec/spec_helper
homebrew/bundle/spec/bundle/cask_installer_spec
homebrew/bundle/spec/bundle/skipper_spec
homebrew/bundle/spec/bundle/brewfile_spec
homebrew/bundle/spec/bundle/cask_dumper_spec
homebrew/bundle/spec/bundle/mac_app_store_installer_spec
homebrew/bundle/spec/bundle/tap_dumper_spec
homebrew/bundle/spec/bundle/whalebrew_installer_spec
homebrew/bundle/spec/bundle/dumper_spec
homebrew/bundle/spec/bundle/dsl_spec
homebrew/bundle/spec/bundle/tap_installer_spec
homebrew/bundle/spec/bundle/brew_installer_spec
homebrew/bundle/spec/bundle/commands/check_command_spec
homebrew/bundle/spec/bundle/commands/exec_command_spec
homebrew/bundle/spec/bundle/commands/list_command_spec
homebrew/bundle/spec/bundle/commands/dump_command_spec
homebrew/bundle/spec/bundle/commands/install_command_spec
homebrew/bundle/spec/bundle/commands/cleanup_command_spec
homebrew/bundle/spec/bundle/brew_dumper_spec
homebrew/bundle/spec/bundle/mac_app_store_dumper_spec
homebrew/bundle/spec/bundle/locker_spec
homebrew/bundle/spec/bundle/brew_services_spec
homebrew/bundle/spec/bundle/whalebrew_dumper_spec
homebrew/bundle/spec/bundle_utils_spec
homebrew/bundle/spec/stub/extend/ENV
homebrew/bundle/spec/stub/exceptions
homebrew/bundle/spec/stub/utils/formatter
homebrew/bundle/spec/stub/tap
homebrew/bundle/spec/stub/utils
homebrew/bundle/spec/stub/os
homebrew/bundle/spec/stub/formulary
homebrew/bundle/spec/stub/development_tools
homebrew/bundle/spec/stub/global
homebrew/bundle/spec/stub/formula
homebrew/bundle/lib/bundle/extend/os/skipper
homebrew/bundle/lib/bundle/extend/os/linux/skipper
homebrew/bundle/lib/bundle/extend/os/linux/bundle
homebrew/bundle/lib/bundle/extend/os/bundle
homebrew/bundle/lib/bundle/checker
homebrew/bundle/lib/bundle/brew_installer
homebrew/bundle/lib/bundle/brew_services
homebrew/bundle/lib/bundle/lister
homebrew/bundle/lib/bundle/brew_service_checker
homebrew/bundle/lib/bundle/mac_app_store_dumper
homebrew/bundle/lib/bundle/brewfile
homebrew/bundle/lib/bundle/cask_installer
homebrew/bundle/lib/bundle/brew_checker
homebrew/bundle/lib/bundle/cask_dumper
homebrew/bundle/lib/bundle/installer
homebrew/bundle/lib/bundle/mac_app_store_installer
homebrew/bundle/lib/bundle/tap_dumper
homebrew/bundle/lib/bundle/whalebrew_dumper
homebrew/bundle/lib/bundle/dumper
homebrew/bundle/lib/bundle/cask_checker
homebrew/bundle/lib/bundle/skipper
homebrew/bundle/lib/bundle/tap_installer
homebrew/bundle/lib/bundle/locker
homebrew/bundle/lib/bundle/commands/install
homebrew/bundle/lib/bundle/commands/dump
homebrew/bundle/lib/bundle/commands/cleanup
homebrew/bundle/lib/bundle/commands/exec
homebrew/bundle/lib/bundle/commands/check
homebrew/bundle/lib/bundle/commands/list
homebrew/bundle/lib/bundle/tap_checker
homebrew/bundle/lib/bundle/bundle
homebrew/bundle/lib/bundle/whalebrew_installer
homebrew/bundle/lib/bundle/dsl
homebrew/bundle/lib/bundle/mac_app_store_checker
homebrew/bundle/lib/bundle/brew_dumper
homebrew/bundle/lib/bundle
homebrew/services/spec/spec_helper
homebrew/services/spec/homebrew/services_cli_spec
homebrew/services/spec/homebrew/service_spec
homebrew/services/spec/stub/exceptions
homebrew/services/lib/services_cli
homebrew/services/lib/service
```
</details>

Reply:
Sorry, I got kind of swamped with other stuff.  Changes made.  @maxim-belkin Thanks for the separator tip, I changed the regexes accordingly.  However, using extended regular expressions is actually fairly expensive, so leaving the command split into 3 sed commands, but only one using the `-E` flag, is slightly faster (by like 10ms, you have to run a bunch of tests and average them to even measure it).  

@MikeMcQuaid I made the changes matching your style.  It certainly seems less hacky that way.  Thanks for the help, and sorry I took so long to get around to this.

Reply:
Hmm, it looks like it is failing because brew man changed?  I'm not sure what to do, maybe someone needs to manually approve my documentation for the formulae.sh command?